### PR TITLE
mgr/cephadm: autotune osd_memory_target

### PR DESCRIPTION
**Pull request by liewegas**
- report both limits and current/recent memory usage in 'ceph orch ps' result
- `osd_memory_target_autotune` bool (default: false) will allow mgr to tune the memory
- rechecked every 10m

### CODE REVIEW

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -4429,6 +4450,22 @@ def list_daemons(ctx, detail=True, legacy_dir=None):
...
+        verbosity=CallVerbosity.DEBUG
+    )
+    seen_memusage_cid_len = 0
+    if not code:
+        for line in out.splitlines():
+            (cid, usage) = line.split(',')
+            (used, limit) = usage.split(' / ')
+            seen_memusage[cid] = with_units_to_int(used)
+            if not seen_memusage_cid_len:
+                seen_memusage_cid_len = len(cid)
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:4500]:
```suggestion
seen_memusage = get_container_memory_uage()
```

I think this deserves it's own function.

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -4619,6 +4652,16 @@ def list_daemons(ctx, detail=True, legacy_dir=None):
...
                                 pass
+
+                        # unit.meta?
+                        mfile = os.path.join(data_dir, fsid, j, 'unit.meta') # type: ignore
+                        try:
+                            with open(mfile, 'r') as f:
+                                meta = json.loads(f.read())
+                                val.update(meta)
+                        except IOError:
+                            pass
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:4663]:
```suggestion
                            logger.info(f'unable to read {mfile}: {str(e)}. ignoring.')
```

File: src/pybind/mgr/orchestrator/module.py [hunk 3]
```diff
@@ -663,6 +674,8 @@ def ukn(s: Optional[str]) -> str:
                     status,
                     nice_delta(now, s.last_refresh, ' ago'),
                     nice_delta(now, s.created),
+                    nice_bytes(s.memory_usage),
+                    nice_bytes(s.memory_request) + ' (' + nice_bytes(s.memory_limit) + ')',
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator/module.py:678]: Paul added a `wide` flage for `device ls` due to the table being too wide to fit on usual screens:

https://github.com/ceph/ceph/blob/316b4ad614a1f662bcf6a4a336b1deb592aeea93/src/pybind/mgr/orchestrator/module.py#L455-L465

wdyt?

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 4]
```diff
@@ -51,6 +53,7 @@ def __init__(self, host: str, daemon_id: str,
 
         # would be great to have the spec always available:
         self.spec: Optional[ServiceSpecs] = spec
+        self.resources: Optional[ResourcesSpec] = resources
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:56]: No need to add that here.  This is going to be removed in:

https://github.com/ceph/ceph/pull/39495/files#diff-95f10b8f6c554f23d766c56a7549b7f05291ff95e6d96a106e6980e49439d62eL52-L54

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 5]
```diff
@@ -107,15 +119,16 @@ def __init__(self, mgr: "CephadmOrchestrator"):
...
                          daemon_type: Optional[str] = None,) -> CephadmDaemonSpec:
         return CephadmDaemonSpec(
             host=host,
             daemon_id=daemon_id,
             spec=spec,
-            network=netowrk,
-            daemon_type=daemon_type
+            network=network,
+            daemon_type=daemon_type,
+            resources=spec.resources,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:131]: I would like to keep the resources within the spec. :-1: for separating it, except we **really** want to deploy individual daemons of a services with specific memory limits.

**Review by sebastian-philipp — DISMISSED** [review]

File: doc/cephadm/osd.rst [hunk 6]
```diff
@@ -363,6 +363,52 @@ Example command:
...
+memory with other services, cephadm can automatically adjust the per-OSD
+memory consumption based on the total amount of RAM and the number of deployed
+OSDs.
+
+This option is enabled globally with::
+
+  ceph config set osd osd_memory_target_autotune true
+
+Cephadm will start with a fraction
+(:confval:`mgr/cephadm/autotune_memory_target_ratio`, which defaults to
```
Comment (tchaikov) [inline doc/cephadm/osd.rst:383]: i am afraid that we don't yet have module options extracted automatically at this moment. i can implement it.

Comment (tchaikov) [reply doc/cephadm/osd.rst:383]: before that we have to define an option like https://github.com/ceph/ceph/blob/master/doc/rados/configuration/general-config-ref.rst . see `max_open_files`.

Comment (liewegas) [reply doc/cephadm/osd.rst]: ok, leaving it off for now

**Comment by liewegas** [comment]
``orch ps`` snapshot from one run:
```
2021-05-06T19:41:01.092 INFO:teuthology.orchestra.run.smithi016.stdout:NAME                            HOST       PORTS        STATUS         REFRESHED  AGE  MEM USED  TARGET  VERSION                IMAGE ID      CONTAINER ID
2021-05-06T19:41:01.092 INFO:teuthology.orchestra.run.smithi016.stdout:alertmanager.smithi016          smithi016  *:9093,9094  running (7m)      8s ago   8m     10.9M       -  0.20.0                 0881eb8f169f  be69c0279a25
2021-05-06T19:41:01.093 INFO:teuthology.orchestra.run.smithi016.stdout:cephfs-mirror.smithi016.xiylqc  smithi016               running (16s)     8s ago  16s     12.9M       -  17.0.0-3900-gff422198  3fd33d75a753  2cb5deb00482
2021-05-06T19:41:01.093 INFO:teuthology.orchestra.run.smithi016.stdout:cephfs-mirror.smithi023.mnlrbi  smithi023               running (14s)     9s ago  14s     13.0M       -  17.0.0-3900-gff422198  3fd33d75a753  12b38a1ab012
2021-05-06T19:41:01.093 INFO:teuthology.orchestra.run.smithi016.stdout:crash.smithi016                 smithi016               running (8m)      8s ago   8m     7312k       -  17.0.0-3900-gff422198  3fd33d75a753  66944c4232fe
2021-05-06T19:41:01.093 INFO:teuthology.orchestra.run.smithi016.stdout:crash.smithi023                 smithi023               running (8m)      9s ago   8m     9.83M       -  17.0.0-3900-gff422198  3fd33d75a753  028f8f1872c6
2021-05-06T19:41:01.094 INFO:teuthology.orchestra.run.smithi016.stdout:grafana.smithi016               smithi016  *:3000       running (7m)      8s ago   8m     23.9M       -  6.7.4                  ae5c36c3d3cd  c05d3c4e24e4
2021-05-06T19:41:01.094 INFO:teuthology.orchestra.run.smithi016.stdout:mgr.smithi016.bxbemg            smithi016  *:9283       running (10m)     8s ago  10m      421M       -  17.0.0-3900-gff422198  3fd33d75a753  ac679947d772
2021-05-06T19:41:01.094 INFO:teuthology.orchestra.run.smithi016.stdout:mgr.smithi023.vfzaup            smithi023  *:8443,9283  running (7m)      9s ago   7m      364M       -  17.0.0-3900-gff422198  3fd33d75a753  8bf0f5d2dfaa
2021-05-06T19:41:01.094 INFO:teuthology.orchestra.run.smithi016.stdout:mon.smithi016                   smithi016               running (10m)     8s ago  10m     49.5M   2048M  17.0.0-3900-gff422198  3fd33d75a753  8a1d0f1dfedf
2021-05-06T19:41:01.094 INFO:teuthology.orchestra.run.smithi016.stdout:mon.smithi023                   smithi023               running (7m)      9s ago   7m     37.7M   2048M  17.0.0-3900-gff422198  3fd33d75a753  d917a7b66ce0
2021-05-06T19:41:01.095 INFO:teuthology.orchestra.run.smithi016.stdout:node-exporter.smithi016         smithi016  *:9100       running (8m)      8s ago   8m     11.6M       -  0.18.1                 e5a616e4b9cf  411e58182ae4
2021-05-06T19:41:01.095 INFO:teuthology.orchestra.run.smithi016.stdout:node-exporter.smithi023         smithi023  *:9100       running (7m)      9s ago   7m     11.3M       -  0.18.1                 e5a616e4b9cf  e0686e2304b5
2021-05-06T19:41:01.095 INFO:teuthology.orchestra.run.smithi016.stdout:osd.0                           smithi016               running (7m)      8s ago   7m     37.2M   3265M  17.0.0-3900-gff422198  3fd33d75a753  fe5e9a5fde08
2021-05-06T19:41:01.095 INFO:teuthology.orchestra.run.smithi016.stdout:osd.1                           smithi016               running (6m)      8s ago   6m     39.3M   3265M  17.0.0-3900-gff422198  3fd33d75a753  77140ddb9679
2021-05-06T19:41:01.096 INFO:teuthology.orchestra.run.smithi016.stdout:osd.2                           smithi016               running (6m)      8s ago   6m     36.7M   3265M  17.0.0-3900-gff422198  3fd33d75a753  1a658f71acd9
2021-05-06T19:41:01.096 INFO:teuthology.orchestra.run.smithi016.stdout:osd.3                           smithi016               running (5m)      8s ago   5m     36.2M   3265M  17.0.0-3900-gff422198  3fd33d75a753  5e23bf9c85ae
2021-05-06T19:41:01.096 INFO:teuthology.orchestra.run.smithi016.stdout:osd.4                           smithi023               running (5m)      9s ago   5m     37.4M   4032M  17.0.0-3900-gff422198  3fd33d75a753  61fafce45aa2
2021-05-06T19:41:01.096 INFO:teuthology.orchestra.run.smithi016.stdout:osd.5                           smithi023               running (4m)      9s ago   4m     36.2M   4032M  17.0.0-3900-gff422198  3fd33d75a753  5ad337c877c2
2021-05-06T19:41:01.096 INFO:teuthology.orchestra.run.smithi016.stdout:osd.6                           smithi023               running (4m)      9s ago   4m     35.5M   4032M  17.0.0-3900-gff422198  3fd33d75a753  bf09684a951a
2021-05-06T19:41:01.097 INFO:teuthology.orchestra.run.smithi016.stdout:osd.7                           smithi023               running (3m)      9s ago   4m     36.2M   4032M  17.0.0-3900-gff422198  3fd33d75a753  e3c190706c82
2021-05-06T19:41:01.097 INFO:teuthology.orchestra.run.smithi016.stdout:prometheus.smithi016            smithi016  *:9095       running (7m)      8s ago   8m     29.8M       -  2.18.1                 de242295e225  92cb43f6caca
2021-05-06T19:41:01.097 INFO:teuthology.orchestra.run.smithi016.stdout:rbd-mirror.smithi016.ukrnqx     smithi016               running (26s)     8s ago  26s     12.9M       -  17.0.0-3900-gff422198  3fd33d75a753  1b671a7a2353
2021-05-06T19:41:01.097 INFO:teuthology.orchestra.run.smithi016.stdout:rbd-mirror.smithi023.jcdjbx     smithi023               running (30s)     9s ago  31s     12.8M       -  17.0.0-3900-gff422198  3fd33d75a753  db5cdbfd06da
```

**Comment by liewegas** [comment]
https://pulpito.ceph.com/sage-2021-05-07_13:48:47-rados:cephadm-wip-sage-testing-2021-05-07-0800-distro-basic-smithi/

**Pull request merged**